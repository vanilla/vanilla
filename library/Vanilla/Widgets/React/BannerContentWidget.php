<?php
/**
 * @author Mihran Abrahamian <mihran.abrahamian@vanillaforums.com>
 * @copyright 2009-2022 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Widgets\React;

use Garden\Schema\Schema;
use Vanilla\Utility\SchemaUtils;
use Vanilla\Widgets\React\BannerFullWidget as ReactBannerFullWidget;
use Gdn_Upload;

/**
 * Class BannerContentWidget
 */
class BannerContentWidget extends ReactBannerFullWidget {

    /**
     * @inheritDoc
     */
    public static function getWidgetName(): string {
        return "Content Banner";
    }

    /**
     * @inheritDoc
     */
    public static function getWidgetID(): string {
        return "banner.content";
    }

    /**
     * @return string
     */
    public static function getWidgetIconPath(): string {
        return "/applications/dashboard/design/images/widgetIcons/contentbanner.svg";
    }

    /**
     * @inheritDoc
     */
    public function getProps(): ?array {
        $parentProps = parent::getProps();
        $categoryID = $this->props['categoryID'] ?? null;
        $iconUrl = $this->categoryModel->getCategoryFieldRecursive($categoryID, 'Photo', null);
        $iconUrl = $iconUrl ? Gdn_Upload::url($iconUrl) : null;
        $props = array_merge($parentProps, [
            'isContentBanner' => true,
            'iconImage' => $iconUrl,
        ], $this->props);
        return $props;
    }

    /**
     * @inheritDoc
     */
    public static function getWidgetSchema(): Schema {
        return SchemaUtils::composeSchemas(
            parent::getWidgetSchema(),
            Schema::parse([
                'iconImage:s?' => 'URL for the icon image.'
            ])
        );
    }
}
