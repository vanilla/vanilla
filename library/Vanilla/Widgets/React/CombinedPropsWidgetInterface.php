<?php
/**
 * @author Adam Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2021 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Widgets\React;

/**
 * Interface for something that can tag a group of props.
 */
interface CombinedPropsWidgetInterface
{
    /**
     * Set the props of a react component.
     *
     * @param array<string, mixed> $props The props
     * @param bool $merge Whether or not we should merge with existing props.
     */
    public function setProps(array $props, bool $merge = false);
}
