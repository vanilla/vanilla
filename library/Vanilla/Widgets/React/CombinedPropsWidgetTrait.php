<?php
/**
 * @author Adam Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2021 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Widgets\React;

/**
 * Trait implementing CombinedPropsWidgetInterface.
 */
trait CombinedPropsWidgetTrait
{
    /** @var array */
    protected $props = [];

    /**
     * @inheritdoc
     */
    public function setProps(array $props, bool $merge = false)
    {
        if ($merge) {
            $this->props = array_replace_recursive($this->props, $props);
        } else {
            $this->props = $props;
        }
    }

    /**
     * @inheritdoc
     */
    public function getProps(): ?array
    {
        return $this->props;
    }
}
