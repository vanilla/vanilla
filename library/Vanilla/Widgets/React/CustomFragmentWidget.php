<?php
/**
 * @author Adam Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2025 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Widgets\React;

use Garden\Schema\Schema;
use Vanilla\Dashboard\Models\FragmentModel;
use Vanilla\Layout\Resolvers\CustomFragmentResolver;

/**
 * Widget for custom fragments. For usage with {@link CustomFragmentResolver}.
 *
 * Instead of having a PHP declared widget, the widget definition is based off of a row in GDN_fragment.
 *
 * @see FragmentModel
 */
class CustomFragmentWidget extends ReactWidget
{
    use AllSectionTrait;

    /** @var array Fragment row from GDN_fragment */
    protected array $fragmentRow = [];

    /**
     * @param array $fragmentRow
     */
    public function setFragmentRow(array $fragmentRow): void
    {
        $this->fragmentRow = $fragmentRow;
    }

    /**
     * @inheritdoc
     */
    public function renderSeoHtml(array $props): ?string
    {
        // Custom no HTML for it.
        return "";
    }

    /**
     * @inheritdoc
     */
    public function getProps(): ?array
    {
        $props = $this->props;

        // Add on fragment info
        $props["fragmentImpl"] = $this->fragmentRow;
        return $props;
    }

    /**
     * @inheritdoc
     */
    public static function getComponentName(): string
    {
        return "CustomFragmentWidget";
    }

    /**
     * @inheritdoc
     */
    public static function getWidgetIconPath(): ?string
    {
        return "/applications/dashboard/design/images/widgetIcons/customhtml.svg";
    }

    /**
     * This is a custom fragment and we don't know the schema statically.
     *
     * It does get joined into the catalog through {@link CustomFragmentResolver} though.
     *
     * @return Schema
     */
    public static function getWidgetSchema(): Schema
    {
        return new Schema(["type" => "object"]);
    }

    /**
     * @inheritdoc
     */
    public static function getWidgetName(): string
    {
        return "Custom Fragment";
    }

    /**
     * @inheritdoc
     */
    public static function getWidgetGroup(): string
    {
        return "Custom";
    }

    /**
     * @inheritdoc
     */
    public static function getWidgetID(): string
    {
        return "custom-fragment";
    }
}
