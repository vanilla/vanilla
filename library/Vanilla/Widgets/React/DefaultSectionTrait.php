<?php
/**
 * @author Adam Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2022 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Widgets\React;

use Vanilla\Layout\Section\SectionOneColumn;
use Vanilla\Layout\Section\SectionThreeColumns;
use Vanilla\Layout\Section\SectionThreeColumnsEven;
use Vanilla\Layout\Section\SectionTwoColumns;
use Vanilla\Layout\Section\SectionTwoColumnsEven;

/**
 * All placement in all sections except for full width.
 */
trait DefaultSectionTrait
{
    /**
     * Allows all sections except for full width.
     *
     */
    public static function getAllowedSectionIDs(): array
    {
        return [
            SectionOneColumn::getWidgetID(),
            SectionTwoColumns::getWidgetID(),
            SectionThreeColumns::getWidgetID(),
            SectionThreeColumnsEven::getWidgetID(),
            SectionTwoColumnsEven::getWidgetID(),
        ];
    }
}
