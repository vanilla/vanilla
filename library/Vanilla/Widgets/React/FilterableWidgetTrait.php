<?php
/**
 * @author David Barbier <dbarbier@higherlogic.com>
 * @copyright 2009-2023 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Widgets\React;

use Garden\Container\ContainerException;
use Garden\Container\NotFoundException;
use Garden\EventManager;
use Garden\Schema\Schema;
use Gdn;
use Vanilla\Forms\ApiFormChoices;
use Vanilla\Forms\FieldMatchConditional;
use Vanilla\Forms\FormOptions;
use Vanilla\Forms\SchemaForm;
use Vanilla\Forms\StaticFormChoices;
use Vanilla\Forum\Controllers\Api\DiscussionsApiIndexSchema;
use Vanilla\Site\DefaultSiteSection;
use Vanilla\Site\SiteSectionModel;
use Vanilla\Utility\ArrayUtils;

/**
 * Trait for filterable widget.
 */
trait FilterableWidgetTrait
{
    /**
     * Get filters schema.
     *
     * @param string[] $usedFilterTypes
     * @param bool|array $followedFilter
     * @param array $extraOptions Possible keys with boolean values: `hasSubcommunitySubTypeOptions`, `hasCategorySubTypeOptions`, `hasGroupSubTypeOptions`.
     * @return Schema
     * @throws ContainerException
     * @throws NotFoundException
     */
    public static function filterTypeSchema(
        array $usedFilterTypes = ["subcommunity", "category", "group", "featured", "none"],
        $followedFilter = false,
        array $extraOptions = []
    ): Schema {
        $fullSchema = [];
        $filterTypes = [];

        $eventManager = Gdn::getContainer()->get(EventManager::class);
        $filterTypes = $eventManager->fireFilter("filterableWidgetTrait_beforeGetFilterTypes", $filterTypes);

        // Every available filter types
        $filterTypes = $filterTypes + [
            "category" => t("Category"),
            "group" => t("Group"),
            "featured" => t("Featured"),
            "none" => t("None"),
        ];

        $filterTypes = ArrayUtils::pluck($filterTypes, $usedFilterTypes);

        // Filter Type
        $fullSchema += [
            "filter" => [
                "type" => "string",
                "default" => in_array("none", $usedFilterTypes) ? "none" : array_key_first($filterTypes),
                "x-control" => SchemaForm::dropDown(
                    new FormOptions(t("Filter By"), t("How the content is going to be filtered.")),
                    new StaticFormChoices($filterTypes)
                ),
            ],
        ];

        // Sub Filter Type for Subcommunity
        if (isset($filterTypes["subcommunity"])) {
            $fullSchema += self::subcommunityFilterTypeSchema($extraOptions["hasSubcommunitySubTypeOptions"] ?? true);
        }

        // Sub Filter Type for category
        if (isset($filterTypes["category"])) {
            $fullSchema += self::categoryFilterTypeSchema($extraOptions["hasCategorySubTypeOptions"] ?? true);
        }

        // Sub Filter Type for group
        if (isset($filterTypes["group"])) {
            $fullSchema += self::groupFilterTypeSchema($extraOptions["hasGroupSubTypeOptions"] ?? true);
        }

        // Sub Filter Type for featured
        if (isset($filterTypes["featured"])) {
            $fullSchema += self::featuredCategoriesFilterTypeSchema();
        }

        $conditionalFollowedFilter = null;
        if (is_array($followedFilter)) {
            $conditionalFollowedFilter = $followedFilter;
        }
        if ($followedFilter || is_array($followedFilter)) {
            $fullSchema += self::followedFilterTypeSchema($conditionalFollowedFilter);
        }

        return Schema::parse($fullSchema);
    }

    /**
     * Get all site-sections form options.
     *
     * @return array
     * @throws ContainerException
     * @throws NotFoundException
     */
    protected static function getSiteSectionFormOptions(): array
    {
        /** @var SiteSectionModel $siteSectionModel */
        $siteSectionModel = Gdn::getContainer()->get(SiteSectionModel::class);
        $siteSections = $siteSectionModel->getAll();

        // If there's only one site-section (default) then we don't
        // need to build the choices.
        if (count($siteSections) === 1) {
            return [];
        }

        $siteSectionFormChoices = [];
        foreach ($siteSections as $siteSection) {
            $id = $siteSection->getSectionID();
            $name = $siteSection->getSectionName();

            if ($id !== DefaultSiteSection::DEFAULT_ID) {
                $siteSectionFormChoices[$id] = $name;
            }
        }

        return $siteSectionFormChoices;
    }

    /**
     * Returns schema chunks for subcommunity filter type.
     *
     * @return array|array[]
     * @throws ContainerException
     * @throws NotFoundException
     */
    public static function subcommunityFilterTypeSchema($hasSubTypeOption = true): array
    {
        $schema = [];

        if ($hasSubTypeOption) {
            $schema += [
                "filterSubcommunitySubType" => [
                    "type" => "string",
                    "default" => "set",
                    "x-control" => SchemaForm::radio(
                        new FormOptions("", t("How the content is going to be filtered.")),
                        new StaticFormChoices([
                            "contextual" => t("Contextual"),
                            "set" => t("Set"),
                        ]),
                        new FieldMatchConditional(
                            "apiParams",
                            Schema::parse(["filter" => ["const" => "subcommunity"]])
                        ),
                        [
                            "contextual" => t(
                                "Content shown is filtered dynamically based on the subcommunity the user is viewing."
                            ),
                        ]
                    ),
                ],
            ];
        }

        // Subcommunity / Site section
        $schema += [
            "siteSectionID" => [
                "type" => ["string", "null"],
                "default" => null,
                "x-control" => SchemaForm::dropDown(
                    new FormOptions(t("Subcommunity"), t("Display records from this subcommunity")),
                    new StaticFormChoices(self::getSiteSectionFormOptions()),
                    new FieldMatchConditional(
                        "apiParams",
                        Schema::parse(
                            $hasSubTypeOption
                                ? [
                                    "filter" => ["const" => "subcommunity"],
                                    "filterSubcommunitySubType" => ["const" => "set"],
                                ]
                                : ["filter" => ["const" => "subcommunity"]]
                        )
                    )
                ),
            ],
        ];

        return $schema;
    }

    /**
     * Returns schema chunks for category filter type.
     *
     * @param $hasSubTypeOption
     * @return array|array[]
     */
    private static function categoryFilterTypeSchema($hasSubTypeOption = true): array
    {
        $schema = [];

        if ($hasSubTypeOption) {
            $schema += [
                "filterCategorySubType" => [
                    "type" => "string",
                    "default" => "set",
                    "x-control" => SchemaForm::radio(
                        new FormOptions("", t("How the content is going to be filtered.")),
                        new StaticFormChoices([
                            "contextual" => t("Contextual"),
                            "set" => t("Set"),
                        ]),
                        new FieldMatchConditional("apiParams", Schema::parse(["filter" => ["const" => "category"]])),
                        [
                            "contextual" => t(
                                "Content shown is filtered dynamically based on the category the user is viewing."
                            ),
                        ]
                    ),
                ],
            ];
        }

        // Category (Set)
        $schema += [
            "categoryID" => [
                "type" => ["integer", "array", "null"],
                "default" => null,
                "x-control" => DiscussionsApiIndexSchema::getCategoryIDFormOptions(
                    new FieldMatchConditional(
                        "apiParams",
                        Schema::parse(
                            $hasSubTypeOption
                                ? [
                                    "filter" => ["const" => "category"],
                                    "filterCategorySubType" => ["const" => "set"],
                                ]
                                : ["filter" => ["const" => "category"]]
                        )
                    )
                ),
            ],
            "includeChildCategories?" => [
                "type" => "boolean",
                "default" => false,
                "x-control" => SchemaForm::toggle(
                    new FormOptions("Include content from subcategories"),
                    new FieldMatchConditional("apiParams", Schema::parse(["filter" => ["const" => "category"]]))
                ),
            ],
        ];

        return $schema;
    }

    /**
     * Returns schema chunks for group filter type.
     *
     * @param $hasSubTypeOption
     * @return array|array[]
     */
    private static function groupFilterTypeSchema($hasSubTypeOption = true): array
    {
        $schema = [];

        if ($hasSubTypeOption) {
            $schema += [
                "filterGroupSubType?" => [
                    "type" => "string",
                    "default" => "set",
                    "x-control" => SchemaForm::radio(
                        new FormOptions("", t("How the content is going to be filtered.")),
                        new StaticFormChoices([
                            "contextual" => t("Contextual"),
                            "set" => t("Set"),
                        ]),
                        new FieldMatchConditional("apiParams", Schema::parse(["filter" => ["const" => "group"]])),
                        [
                            "contextual" => t(
                                "Content shown is filtered dynamically based on the group the user is viewing."
                            ),
                        ]
                    ),
                ],
            ];
        }

        // Group (Set)
        $schema += [
            "groupID" => [
                "type" => ["integer", "null"],
                "default" => null,
                "x-control" => SchemaForm::dropDown(
                    new FormOptions("Group", "Choose a group"),
                    new ApiFormChoices("/api/v2/groups", "/api/v2/groups/%s", "groupID", "name"),
                    new FieldMatchConditional(
                        "apiParams",
                        Schema::parse(
                            $hasSubTypeOption
                                ? [
                                    "filter" => ["const" => "group"],
                                    "filterCategorySubType" => ["const" => "set"],
                                ]
                                : ["filter" => ["const" => "group"]]
                        )
                    )
                ),
            ],
        ];

        return $schema;
    }

    /**
     * Returns schema chunks for featured categories filter type.
     *
     * @return array|array[]
     */
    private static function featuredCategoriesFilterTypeSchema(): array
    {
        $schema = [];

        $schema += [
            "featuredCategoryID?" => [
                "type" => ["array", "null"],
                "default" => null,
                "x-control" => DiscussionsApiIndexSchema::getCategoryIDFormOptions(
                    new FieldMatchConditional("apiParams", Schema::parse(["filter" => ["const" => "featured"]])),
                    true
                ),
            ],
        ];

        return $schema;
    }

    /**
     * Returns schema chunks for followed filter type.
     *
     * @param null|array $conditionalFollowedFilter
     * @return array|array[]
     */
    private static function followedFilterTypeSchema($conditionalFollowedFilter = null): array
    {
        if ($conditionalFollowedFilter === null) {
            $conditionalFollowedFilter = ["subcommunity", "category", "none"];
        }

        $schema = [];

        $schema += [
            "followed?" => [
                "type" => "boolean",
                "default" => false,
                "x-control" => SchemaForm::checkBox(
                    new FormOptions(t("Only Show Followed Categories"), t("Display content from followed categories")),
                    new FieldMatchConditional(
                        "apiParams",
                        Schema::parse(["filter" => ["enum" => $conditionalFollowedFilter]])
                    )
                ),
            ],
        ];

        return $schema;
    }
}
