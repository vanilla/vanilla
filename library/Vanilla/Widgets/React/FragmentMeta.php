<?php
/**
 * @author Adam Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2025 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Widgets\React;

use Garden\Schema\Schema;

abstract class FragmentMeta implements \JsonSerializable
{
    /**
     * Get the type of the fragment.
     *
     * @return string
     */
    abstract public static function getFragmentType(): string;

    abstract public static function getName(): string;

    /**
     * Get the schema of props passed to the fragment.
     *
     * @return Schema
     */
    abstract public function getPropSchema(): Schema;

    /**
     * @return bool
     */
    public static function isAvailableInStyleguide(): bool
    {
        return false;
    }

    /**
     * @return array
     */
    public function jsonSerialize(): array
    {
        return [
            "fragmentType" => static::getFragmentType(),
            "schema" => $this->getPropSchema(),
            "name" => static::getName(),
        ];
    }
}
