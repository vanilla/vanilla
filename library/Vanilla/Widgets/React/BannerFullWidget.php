<?php
/**
 * @author Mihran Abrahamian <mihran.abrahamian@vanillaforums.com>
 * @copyright 2009-2022 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Widgets\React;

use CategoryModel;
use Garden\Schema\Schema;
use Gdn_Upload;
use Vanilla\Contracts\ConfigurationInterface;
use Vanilla\Widgets\HomeWidgetContainerSchemaTrait;

/**
 * Class BannerFullWidget
 */
class BannerFullWidget implements ReactWidgetInterface, CombinedPropsWidgetInterface {

    use CombinedPropsWidgetTrait;
    use HomeWidgetContainerSchemaTrait;

    /** \CategoryModel */
    protected $categoryModel;
    /** ConfigurationInterface */
    private $config;

    /**
     * DI.
     *
     * @param CategoryModel $categoryModel
     * @param ConfigurationInterface $config
     */
    public function __construct(CategoryModel $categoryModel, ConfigurationInterface $config) {
        $this->categoryModel = $categoryModel;
        $this->config = $config;
    }

    /**
     * @inheritDoc
     */
    public static function getWidgetName(): string {
        return "Banner";
    }

    /**
     * @inheritDoc
     */
    public static function getWidgetID(): string {
        return "banner.full";
    }

    /**
     * @return string
     */
    public static function getWidgetIconPath(): string {
        return "/applications/dashboard/design/images/widgetIcons/banner.svg";
    }

    /**
     * @inheritDoc
     */
    public static function getWidgetSchema(): Schema {
        return Schema::parse([
            'backgroundImage:s?' => 'URL for the background image.',
            'categoryID:s?' => 'Category ID.',
            'title:s?' => 'Title.',
            'description:s?' => 'Description.',
        ]);
    }

    /**
     * @inheritDoc
     */
    public function getProps(): ?array {
        $categoryID = $this->props['categoryID'] ?? null;

        $bgUrl = $this->categoryModel->getCategoryFieldRecursive(
            $categoryID,
            'BannerImage',
            $this->config->get("Garden.BannerImage", null)
        );
        $bgUrl = $bgUrl ? Gdn_Upload::url($bgUrl) : null;

        return array_merge([
            'backgroundImage' => $bgUrl,
            'options' => [
                'enabled' => true,
            ],
        ], $this->props);
    }

    /**
     * @inheritDoc
     */
    public static function getComponentName(): string {
        return "Banner";
    }
}
