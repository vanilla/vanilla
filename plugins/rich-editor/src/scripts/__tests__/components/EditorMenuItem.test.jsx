import React from "react";
import { shallow } from "enzyme";
import EditorMenuItem from "../../components/EditorMenuItem";

test("matches snapshot", () => {
    const clickHandler = () => {};
    const item = shallow(
        <EditorMenuItem clickHandler={clickHandler} propertyName="bold"/>
    );

    expect(item).toMatchSnapshot();
});

test("click handler is called", () => {
    const mockClickHandler = jest.fn();
    const item = shallow(
        <EditorMenuItem clickHandler={mockClickHandler} propertyName="bold"/>
    );
    item.find(".richEditor-button").simulate("click");
    expect(mockClickHandler.mock.calls.length).toBe(1);
});
