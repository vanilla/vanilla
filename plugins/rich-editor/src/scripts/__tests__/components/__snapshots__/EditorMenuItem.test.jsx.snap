// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditorMenuItem
    clickHandler={[Function]}
    propertyName="bold"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <button
        className="richEditor-button"
        onClick={[Function]}
        type="button"
      >
        <bold />
      </button>,
      "className": "richEditor-menuItem",
      "role": "menuitem",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <bold />,
        "className": "richEditor-button",
        "onClick": [Function],
        "type": "button",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "button",
    },
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <button
          className="richEditor-button"
          onClick={[Function]}
          type="button"
        >
          <bold />
        </button>,
        "className": "richEditor-menuItem",
        "role": "menuitem",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <bold />,
          "className": "richEditor-button",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "button",
      },
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
